File: 20/valid_parentheses.js
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/
const main = () => {
  // console.log("=========================================");
  // console.log(isValid("()"));
  // console.log(true);
  // console.log("=========================================");
  // console.log(isValid("()[]{}"));
  // console.log(true);
  // console.log("=========================================");
  // console.log(isValid("(]"));
  // console.log(false);
  console.log("=========================================");
  console.log(isValid("aa"));
  console.log(false);
  console.log("=========================================");
  // console.log(isValid("([)]"));
  // console.log(false);
  // console.log("=========================================");
  // console.log(isValid("{[]}"));
  // console.log(true);
  // console.log("=========================================");
};

var isValid = function(s) {
  let stack = [];
  let count = 0;
  for(let i = 0; i < s.length; i++) {
    console.log(i);
    switch (stack[stack.length - 1]) {
      case "(":
        if(s[i] === ")") {
          stack.pop();
        }
        break;
      case "{":
        if(s[i] === "}") {
          stack.pop();
        }
        break;
      case "[":
        if(s[i] === "]") {
          stack.pop();
        }
        break;
      case undefined:
        console.log("undefined");
      wow:
        console.log("wow!");
        // break wow;
        console.log("more!");

      default:
        console.log("push");
        stack.push(s[i]);
        break;
    }
    console.log(stack);
  }
  return !stack.length;
};

main();

File: 20/valid_parentheses.go
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/
package main

import (
	"github.com/k0kubun/pp"
)

func main() {
	pp.Println("=========================================")
	pp.Println(isValid("()"))
	pp.Println(true)
	pp.Println("=========================================")
	pp.Println(isValid("()[]{}"))
	pp.Println(true)
	pp.Println("=========================================")
	pp.Println(isValid("(]"))
	pp.Println(false)
	pp.Println("=========================================")
	pp.Println(isValid("([)]"))
	pp.Println(false)
	pp.Println("=========================================")
	pp.Println(isValid("{[]}"))
	pp.Println(true)
	pp.Println("=========================================")
}

/**
 * [description]
 * - Time Complexity: O(X)
 * - Space Complexity: O(X)
 */

var pairs = map[rune]rune{
	')': '(',
	'}': '{',
	']': '[',
}

// func isValid(s string) bool {
// 	stack := []rune{}
// 	for _, r := range s {
// 		if pair, ok := pairs[r]; ok {
// 			if len(stack) == 0 || stack[len(stack)-1] != pair {
// 				return false
// 			}
// 			stack = stack[:len(stack)-1] // pop
// 		} else {
// 			stack = append(stack, r) // push
// 		}
// 	}
// 	return len(stack) == 0
// }

func isValid(s string) bool {
	stack := []rune{}
	var count int
	for _, b := range s {
		switch b {
		case '(', '{', '[':
			stack = append(stack, b) // push
			goto label2
		label:
			pp.Println("debugging...1")
			count++
			if count > 3 {
				return false
			}
		label2:
			pp.Println("debugging...2")
			goto label
			break
		default:
			var pair rune
			switch b {
			case ')':
				pair = '('
			case '}':
				pair = '{'
			case ']':
				pair = '['
			}
			if len(stack) == 0 {
				return false
			}
			lastRune := stack[len(stack)-1]
			if lastRune != pair {
				return false
			}
			stack = stack[:len(stack)-1] // pop
		}
	}
	return len(stack) == 0
}

// func isValid(s string) bool {
// 	stack := []rune{}
// 	for _, b := range s {
// 		if b == '(' || b == '{' || b == '[' {
// 			stack = append(stack, b) // push
// 		} else {
// 			var pair rune
// 			if b == ')' {
// 				pair = '('
// 			} else if b == '}' {
// 				pair = '{'
// 			} else if b == ']' {
// 				pair = '['
// 			}
// 			if len(stack) == 0 {
// 				return false
// 			}
// 			lastRune := stack[len(stack)-1]
// 			if lastRune != pair {
// 				return false
// 			}
// 			stack = stack[:len(stack)-1] // pop
// 		}
// 	}
// 	return len(stack) == 0
// }
